src/3rd_party_bundle_add.c

```diff
@@ -0,0 +1,105 @@
+ *
+ *   This program is free software: you can redistribute it and/or modify
+ *   it under the terms of the GNU General Public License as published by
+ *   the Free Software Foundation, version 2 or later of the License.
+ *
+ *   This program is distributed in the hope that it will be useful,
+ *   but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *   GNU General Public License for more details.
+ *
+ *   You should have received a copy of the GNU General Public License
+ *   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#define _GNU_SOURCE
+
+#include "swupd.h"
+
+static char **cmdline_bundles;
+
+static void print_help(void)
+{
+	print("Usage:\n");
+	print("   swupd 3rd-party bundle-add [OPTION...] [bundle1 bundle2 (...)]\n\n");
+
+	global_print_help();
+
+	print("Options:\n");
+	print("   --skip-optional         Do not install optional bundles (also-add flag in Manifests)\n");
+	print("   --skip-diskspace-check  Do not check free disk space before adding bundle\n");
+	print("\n");
+}
+
+enum swupd_code third_party_bundle_add_main(int argc, char **argv)
+{
+	int ret = SWUPD_OK;
+
+	/*
+	 * Steps for bundle-add:
+	 *
+	 *  1) load_manifests
+	 *  2) download_packs
+	 *  3) extract_packs
+	 *  4) validate_fullfiles
+	 *  5) download_fullfiles
+	 *  6) extract_fullfiles
+	 *  7) install_files
+	 *  8) run_postupdate_scripts
+	 */
+	const int steps_in_bundleadd = 8;
+
+	if (!bundle_add_parse_options(argc, argv)) {
+		print_help();
+		return SWUPD_INVALID_OPTION;
+	}
+	progress_init_steps("3rd-party-bundle-add", steps_in_bundleadd);
+
+	/* initialize swupd and get current version from OS */
+	ret = swupd_init(SWUPD_ALL);
+	if (ret != 0) {
+		error("Failed updater initialization, exiting now\n");
+		return ret;
+	}
+
+	// TODO(s):
+	// 1) we need to get the content url of the 3rd-party repo from
+	// the repo.ini file instead of being hard coded
+	// 2) when getting the url, if there are many 3rd-party repos, we
+	// need to search for the bundle in all of them
+	// 3) if the bundle is in multiple repos we need to ask the user to
+	// choose one
+	// 4) when getting the current_version of the system we need to get
+	// it from the repo.ini instead of getting it from /usr/lib/os-release
+	// which has the version related to the upstream content
+	// 5) when tracking the installed bundle we need to do one of these two
+	// things:
+	//     a) track it in the same state dir where we track the upstream bundles
+	//        in the target system
+	//     b) track it in the state dir that is specific to the 3rd-party repo
+	//        but then we need to make changes in other functions that look for
+	//        tracked bundles (like bundle-remove) to look for trakced bundles
+	//        in all state dirs
+
+	char *tp_state_dir;
+	string_or_die(&tp_state_dir, "%s/3rd-party/test-repo", globals.state_dir);
```

Comment:
For now I think we can set all those variables the way you are doing. After we have a clear idea on everything we need to set we can split swupd_init() somehow to create a safe way to set/unset all this variables and make sure we won't forget anything.

Reply:
thanks for taking a look at this @otaviobp, I'll continue moving in this direction 😄